tests/unit/test_hnsw_tiered.cpp

```diff
@@ -436,3 +436,465 @@ TYPED_TEST(HNSWTieredIndexTest, insertJobAsyncMulti) {
+            thread_pool[i].join();
+        }
+
+        EXPECT_EQ(tiered_index->index->indexSize(), n) << (isMulti ? "multi" : "single");
+        EXPECT_EQ(tiered_index->index->indexLabelCount(), n_labels)
+            << (isMulti ? "multi" : "single");
+        EXPECT_EQ(tiered_index->flatBuffer->indexSize(), 0) << (isMulti ? "multi" : "single");
+        EXPECT_EQ(tiered_index->labelToInsertJobs.size(), 0) << (isMulti ? "multi" : "single");
+        EXPECT_EQ(successful_searches, n) << (isMulti ? "multi" : "single");
+        EXPECT_EQ(jobQ.size(), 0) << (isMulti ? "multi" : "single");
+
+        // Cleanup.
+        thread_pool.clear();
+    }
+}
+
+TYPED_TEST(HNSWTieredIndexTest, parallelInsertSearch) {
+    size_t dim = 4;
+    size_t k = 10;
+    size_t n = 3000;
+
+    size_t block_size = n / 100;
+
+    // Create TieredHNSW index instance with a mock queue.
+    std::shared_ptr<VecSimAllocator> allocator = VecSimAllocator::newVecsimAllocator();
+    for (auto isMulti : {false, true}) {
+        size_t n_labels = isMulti ? n / 25 : n;
+        HNSWParams params = {
+            .type = TypeParam::get_index_type(),
+            .dim = dim,
+            .metric = VecSimMetric_L2,
+            .multi = isMulti,
+            .blockSize = block_size,
+        };
+        auto jobQ = JobQueue();
+        auto index_ctx = IndexExtCtx();
+        size_t memory_ctx = 0;
+        TieredIndexParams tiered_params = {
+            .jobQueue = &jobQ,
+            .jobQueueCtx = &index_ctx,
+            .submitCb = submit_callback,
+            .memoryCtx = &memory_ctx,
+            .UpdateMemCb = update_mem_callback,
+        };
+        TieredHNSWParams tiered_hnsw_params = {.hnswParams = params, .tieredParams = tiered_params};
+        auto *tiered_index = reinterpret_cast<TieredHNSWIndex<TEST_DATA_T, TEST_DIST_T> *>(
+            HNSWFactory::NewTieredIndex(&tiered_hnsw_params, allocator));
+        // Set the created tiered index in the index external context.
+        index_ctx.index_strong_ref.reset(tiered_index);
+        EXPECT_EQ(index_ctx.index_strong_ref.use_count(), 1);
+
+        // Launch the BG threads loop that takes jobs from the queue and executes them.
+        // Save the number fo tasks done by thread i in the i-th entry.
+        std::vector<size_t> completed_tasks(THREAD_POOL_SIZE, 0);
+        bool run_thread = true;
+        for (size_t i = 0; i < THREAD_POOL_SIZE; i++) {
+            thread_pool.emplace_back(thread_main_loop, std::ref(jobQ), std::ref(run_thread));
+        }
+        std::atomic_int successful_searches(0);
+
+        auto parallel_knn_search = [](AsyncJob *job) {
+            auto *search_job = reinterpret_cast<SearchJobMock *>(job);
+            size_t k = search_job->k;
+            auto query = search_job->query;
+            // In this test we don't care about the results, just that the search doesn't crash
+            // and returns the correct number of valid results.
```

Comment:
why?

Reply:
because it's too hard to predict the results as we don't know the order of execution and what vectors are going to be in the index at the execution time. We have other tests for validating the score or the expected ids, so I think it is better to accept the chaos here and only test the simple stuff (number of results, uniqueness) and not limit the case for better prediction